src/utils/utilities.ts

```diff
@@ -240,3 +244,189 @@ export const calculateTimeLeft = (createdAt: string) => {
+        const voutIndex = psbt.txInputs[index].index;
+        totalInput += tx.outs[voutIndex].value;
+      }
+    });
+
+    outputs.forEach((output) => {
+      totalOutput += output.amount;
+      if (!changeAddress.includes(output.address)) {
+        totalAmount += output.amount;
+      }
+    });
+
+    const receiverAddresses = outputs.map((op) => op.address);
+
+    // Calculate transaction fees
+    const fees = totalInput - totalOutput;
+    const feeRate = (fees / vBytes).toFixed(2);
+    return {
+      senderAddresses: senderAddresses,
+      receiverAddresses: receiverAddresses,
+      fees,
+      sendAmount: totalAmount,
+      currentSignerFromPSBT,
+      signerMatched,
+      changeAddress,
+      txId,
+      vout,
+      feeRate,
+      vBytes,
+    };
+  } catch (error) {
+    console.log('🚀 ~ dataFromPSBT ~ error:', error);
+    throw 'Something went wrong';
+  }
+};
+
+export const estimateVByteFromPSBT = (base64Str: string) => {
+  const psbt = bitcoin.Psbt.fromBase64(base64Str);
+  const unsignedTxHex =
+    psbt.txInputs.length > 0 && psbt.txOutputs.length > 0 ? psbt.__CACHE.__TX.toHex() : null;
+  const tx = bitcoin.Transaction.fromHex(unsignedTxHex);
+
+  // Calculate the base size (without witness data)
+  const baseSize = tx.toBuffer().length;
+
+  // Check if there is witness data and calculate the total size accordingly
+  const totalSize = tx.hasWitnesses() ? tx.virtualSize() : baseSize;
+
+  // Calculate vBytes using the formula
+  return Math.ceil((baseSize * 3 + totalSize) / 4);
+};
+
+export const getTnxDetailsPSBT = (averageTxFees, feeRate: string) => {
+  let estimatedBlocksBeforeConfirmation = 0;
+  let tnxPriority = TxPriority.LOW;
+  if (averageTxFees && averageTxFees[config.NETWORK_TYPE]) {
+    const { high, medium, low } = averageTxFees[config.NETWORK_TYPE];
+    const customFeeRatePerByte = parseInt(feeRate);
+    if (customFeeRatePerByte >= high.feePerByte) {
+      estimatedBlocksBeforeConfirmation = high.estimatedBlocks;
+      tnxPriority = TxPriority.HIGH;
+    } else if (customFeeRatePerByte <= low.feePerByte) {
+      estimatedBlocksBeforeConfirmation = low.estimatedBlocks;
+    } else {
+      estimatedBlocksBeforeConfirmation = medium.estimatedBlocks;
+      tnxPriority = TxPriority.MEDIUM;
+    }
+  }
+  return { estimatedBlocksBeforeConfirmation, tnxPriority };
+};
+export const getChangeAddress = (psbt) => {
+  // Decode the PSBT
+  const decodedPsbt = bitcoin.Psbt.fromBase64(psbt);
+
+  // Gather BIP32 derivation information from the inputs to determine the policy
+  const inputPolicy = decodedPsbt.data.inputs[0].bip32Derivation.map((derivation) => ({
+    pubkey: derivation.pubkey,
+    path: derivation.path,
+    fingerprint: derivation.masterFingerprint.toString('hex'),
+  }));
+
+  // Helper function to check if two arrays of public keys and paths match
+  const compareMultisigPolicy = (policy1, policy2) => {
```

Comment:
This is not enough for secure change verification. Here you merely verify the fingerprints are the same, but those could be manipulated, or simply used with a differnet multisig with the same devices...

What you need to do is derive the addresses from xpubs and ensure you are able to derive both all the inputs and the specific output using the specified derivations and script type.

I think for now it will be fine if we comment this function out to just return an empty array, and decide if we want to do change verification later. And for now just show the change as part of the recipients...

Reply:
We may want to only do change detection for PSBTs which belong to a known wallet, then it'll be a lot simpler also. And for signing txs for unkown vaults we'd just show all outputs

Reply:
removed change calculation from psbt